### PR TITLE
Rewritten OStatus inbound part

### PR DESCRIPTION
**Pull request by annando**
The thread completion in OStatus is completely rewritten. It now acts standard conform and works with Pleroma and Mastodon as well. Additionally Friendica now provides every needed data, so that Pleroma and Mastodon can now better complete threads with Friendica posts.

This morning I had a small issue with Friendica posts in an OStatus thread. This hadn't happen again, so it is hopefully solved.

This should solve issue https://github.com/friendica/friendica/issues/3018 and https://github.com/friendica/friendica/issues/3318.

### CODE REVIEW

File: include/dfrn.php [hunk 1]
```diff
@@ -827,7 +824,7 @@ private static function get_attachment($doc, $root, $item) {
 	 * @return object XML entry object
 	 * @todo Add type-hints
 	 */
-	private static function entry($doc, $type, $item, $owner, $comment = false, $cid = 0) {
+	private static function entry($doc, $type, $item, $owner, $comment = false, $cid = 0, $single = false) {
```
Comment (rabuzarus) [inline include/dfrn.php]: The `$single` parameter does need a description in the doxygen header

File: include/ostatus.php [hunk 2]
```diff
@@ -396,973 +410,631 @@ public static function import($xml,$importer,&$contact, &$hub) {
...
+	 * @brief Create an url out of an uri
 	 *
-	 * @param integer $itemid The id of the item
-	 * @param string $conversation_url The uri of the conversation
+	 * @param string $href URI in the format "parameter1:parameter1:..."
+	 *
+	 * @return string URL in the format http(s)://....
 	 */
-	private static function store_conversation($itemid, $conversation_url) {
+	public static function convert_href($href) {
```
Comment (rabuzarus) [inline include/ostatus.php]: I would rather use camelCase for function naming (`public static function convertHref($href)`)

**Comment by rabuzarus** [comment]
I suggest to convert the old ostatus function names to camelCase naming. e.g. `fetchAuthor()`, `fetchSalmonAuthor()`, `readAttributes()` and so on. So we wouldn't have a mixed up style and could move the file later to the `Friendica\Core\Protocol\OStatus` namespace

File: include/ostatus.php (hunk 2 above)
Comment (annando) [reply include/ostatus.php]: This is an old function. I hadn't rewritten every code line.

**Comment by annando** [comment]
I suggest that we create a separate issue to move all protocol classes to /src/ - and we should do this in the next milestone, okay?

File: include/dfrn.php (hunk 1 above)
Comment (annando) [reply include/dfrn.php]: Yeah, I will do this later today.

**Comment by rabuzarus** [comment]
OK

File: include/dfrn.php (hunk 1 above)
Comment (annando) [reply include/dfrn.php]: Done

**Comment by rabuzarus** [comment]
Without knowing the details about OStatus I would say that the code seems to be fine. Like every time: having more descriptive inline comments which would describe what happens (or why it have to happen) would be nice but I would give the code a try in a productive use

**Pull request merged**